### ISSUE TITLE
Find header compile_commands.json entries from CMake VERIFY_INTERFACE_HEADER_SETS

### ISSUE THREAD

**Issue by benschreiber** [clangd/clangd#1652]
CMake has a [VERIFY_INTERFACE_HEADER_SETS](https://cmake.org/cmake/help/latest/prop_tgt/VERIFY_INTERFACE_HEADER_SETS.html) property. When set, CMake will create a file `<header>.cxx` for every interface header in the target. This cxx file just includes the header to verify that it can be included by itself. When CMAKE_EXPORT_COMPILE_COMMANDS is also on, entries for `<header>.cxx` are added. 

CMake does not usually generate compile_commands entries for headers, forcing clangd to use a heuristic. The command used for `<header>.cxx` would be a good choice to use, and a preference for it should be added to the heuristic.

**Comment by HighCommander4** [comment]
Just to make sure I understand the request:

Are you saying that if the compilation database contains such `<header>.cxx` entries generated by CMake, clangd should use the command from that rather than a different header file?

Or that, in cases where the compilation database does **not** contain such `<header>.cxx` entries, clangd should independently come up with the same command that CMake would if it were to create a `<header>.cxx` file?

**Comment by benschreiber** [comment]
> Are you saying that if the compilation database contains such \<header\>.cxx entries generated by CMake, clangd should use the command from that rather than a different header file?

Yes. But it would be substituting a different cpp file, not a header file. The whole problem is that headers aren't in the compilation database. If clangd doesn't have a good candidate like `foo.cpp` for `foo.h`, then `foo.h.cxx` would be another good choice. clangd is picking a some other random cpp file in my project and it has the wrong flags.

> Or that, in cases where the compilation database does not contain such \<header\>.cxx entries, clangd should independently come up with the same command that CMake would if it were to create a \<header\>.cxx file?

No. I don't think clangd could even do that. CMake has the information about which header paths to include and macros to define.

**Comment by HighCommander4** [comment]
> If clangd doesn't have a good candidate like `foo.cpp` for `foo.h`, then `foo.h.cxx` would be another good choice.

Is `foo.h.cxx` located in the same directory as `foo.h`?

I think clangd should already be preferring `foo.h.cxx` as the source file whose compile command to use for `foo.h` in a case like that, but it may be sensitive to the directory structure, or possibly it's thrown off by the `.h.cxx` double extension. (The code for selecting the preferred file is [here](https://searchfox.org/llvm/rev/48339d0fbbdb2ba4610173ad1e792e9c2272c887/clang/lib/Tooling/InterpolatingCompilationDatabase.cpp#344-364).)

**Comment by benschreiber** [comment]
No, it's not in the same directory. clangd picks a different random file, not `foo.h.cxx`, to infer flags from. When I edit the compilation database entry to remove ".cxx" so that "file" is "foo.h", clangd _does_ pick it. Looking at the code, maybe the extra point for an exact match is what makes the difference. Is there any way to dump or debug this process?

I think `foo.h.cxx` is a better choice than most files, so the heuristic should be modified to prefer it over a random file.

**Comment by HighCommander4** [comment]
> Is there any way to dump or debug this process?

Not that I'm aware of besides attaching a debugger to clangd.

**Comment by HighCommander4** [comment]
To make the issue here concrete, I created a project with the following directory structure:

```
.
├── src
│   ├── include
│   │   ├── interface.h
│   └── unrelated.cxx
└── verify
    └── interface.h.cxx
```

and a `compile_commands.json` that contains an entry for both `.cxx` files.

Indeed, when `interface.h` is opened, clangd prefers to use the command of `unrelated.cxx` instead of `interface.h.cxx`, I guess because `unrelated.cxx` is "closer" in the directory tree.

I think a fix approach could be: if the header file's name is a prefix match for the source's file's name, weight that higher than proximity in the directory tree.
